File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline
class makeStack:
    def __init__(self):
        self.stk=[]
        self.length=0

    def push(self,value):
        self.stk.append(value)
        self.length+=1

    def pop(self):
        if self.length==0:
            return -1
        self.stk.pop()
        self.length-=1
        return 0



n=int(input())
for _ in range(n):
    s=makeStack()
    good=1
    m = str(input())
    for i in m:
        if i=='(':
            s.push(i)
        elif i==')':
            a=s.pop()
            if a==-1:
                good=0
                break
    if good==1 and s.length==0:
        print('YES')
    else:
        print('NO')
    

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static Stack<Character> stk;
    static StringBuilder sb= new StringBuilder();

    static void solution(int n) throws IOException {
        String curr;
        a: for (int i = 0; i < n; i++) {
            stk = new Stack<>();
            curr = br.readLine();
            for (int j = 0; j < curr.length(); j++) {
                if(curr.charAt(j)=='(') stk.push('(');
                else if(!stk.isEmpty()) stk.pop();
                else {
                    sb.append("NO");
                    sb.append("\n");
                    continue a;
                }
            }
            if (stk.isEmpty())
                sb.append("YES");
            else sb.append("NO");
            sb.append("\n");
        }
    }
    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        solution(n);
        System.out.println(sb.toString());
    }
}
